### DISCUSSION TITLE
Seeking color temp adjustment values

### DISCUSSION THREAD

**Discussion by memphis-10** [ImageMagick/ImageMagick#4365] [Digital Image Processing]
I've searched and tested to the best of my ability for a week or two and can't quite figure this out, so here I am.  Thanks in advance for your input!

Situation: 
1- tungsten balanced film shot in daylight without a filter (3200º shot in 5600º)
2- daylight balanced film shot in tungsten without a filter (5600º shot in 3200º)

I'd like to use a set of operators and values to color-correct these images--especially the first example for now (3200º shot in 5600º).  (I can do the corrections in image editing software, but I'm hoping to find an IM method that gets me there--_or at least close_). 

I found a good colortemp script on Fred's page, but I don't need variable temp inputs and am seeking to run the correction in my own IM string during the negative conversion/inversion.

This is the windows script I'm using from the scanner group (Pakon F135 - only runs on XP)

(-level-colors is used to remove the scanner and base color casts.) 

I would like to add the correction after -level-colors in the *negative* stage, therefore I realize the corrections used need to be inverted to run on a negative.  

*.raw to *.tif (whole folder, Windows XP 32 bit, IM 7.0.10 Q16)

magick convert -endian LSB -size 3000x2000+16 -depth 16 -interlace plane %%~nf.rgb -compress none -interlace none -depth 16 -gamma 2.2 %%~nf.tif
del /q %%~nf.rgb magick %%~nf.tif -channel rgb -level-colors "rgb(34,26,29)","rgb(125,108,105)" -negate %%~nf.tif

**Comment by snibgo** [comment]
As I commented on https://github.com/ImageMagick/ImageMagick/discussions/4367 , I have no experience of matching digitised film.

Adjusting colour temperature is usually done by multiplying the linear RGB channels by constants. [EDIT: The multiplication simulates the action of a corrective filter in front of the lens.] There are theoretical formulae for deriving the constants from the temperatures in kelvin. If you have negative values, then the transformation is not a simple multiplication.

To get the linear RGB values, we would need to reverse the characteristic curve of the film. The CC can be derived from a photo of a calibrated step wedge.

**Reply by memphis-10** [reply]
Cheers, thank you.  I realize this is similar to #4367, though somewhat separate in my use.

In this case: simulating a corrective filter in post is exactly what I'm hoping to do.  I do not have a calibrated wedge and am willing to work in positive for now if that's easier.  I'm lost on how to get the values to put in my string.  Is there a known source or method I can start with?

thanks a lot

**Reply by memphis-10** [reply]
Also, I can work with digital for now until I get the hang of it.

**Comment by snibgo** [comment]
If you have a good digital camera, I expect the raw values are linear. That is, values are proportional to the light energy received, possibly plus a constant. If they are plus a constant, you will need to subtract it.

I show how to test whether your camera has [linear raw values](http://im.snibgo.com/lincamraw.htm).

Then, suppose you photograph anything gray, and the linear RGB values are (34%,36%,39%). What multipliers will make those values equal?

One solution is (36/34, 36/36, 36/39) = (1.0588, 1, 0.9231). This takes the green channel as "correct", and balances the other two channels against it.

Another solution is (34/34, 34/36, 34/39) = (1, 0.9444, 0.8718). This takes the minimum channel as correct. It has the advantage that multipliers are all less than or equal to one, so it cannot cause clipping.

Then you can multiply the entire image by those three multipliers to adjust the colour temperature.

**Reply by memphis-10** [reply]
Thank you for taking the time to explain this.  It's a huge help.  My digital camera is a Canon 5D Mk II, which I suspect shoots linear but I'll double-check.  I have chart I can use.  I doubt the colors are true macbeth, but they are close enough for now.

I have been experimenting for a while, some efforts getting really close. 

Can I do the correction on a *.jpg version of the grey patch straight from the camera, or do I need to convert from the *.CR2 file?  And is -evaluate Multiply on the right track? 

This is the "incorrect" balance:
![IMG_5724](https://user-images.githubusercontent.com/88346417/137252606-266e43b6-5096-4aff-a3e2-ccf1f783495e.jpg)
The greys must have a little cast to them, as correcting to those ended up a little green. Using the white background I got closest.

These values: 168.584, 211.436, 244.669

led me to this string:

-channel g -evaluate Multiply 0.7973 -channel b -evaluate Multiply 0.689

which produced this image:

![IMG_5724-mult](https://user-images.githubusercontent.com/88346417/137253423-dc260bf6-31db-4660-8bda-970fed117a46.jpg)

So thanks again for the steps in the right direction.  I've appreciated this a lot.

2 more questions of you have time:
- Is there a better/cleaner way to do the multiplying?
- Will you point me towards matching one grey to another grey?  For example, I have 95,147,187 in the 3rd square in the "incorrect" image and I want to make it:  155,155,133.  I tried the follwoing, but it's not changing the image:

-channel r -evaluate Add 60 -channel g -evaluate Add 8 -channel b -evaluate Subtract 54

Thanks again

**Comment by snibgo** [comment]
> Can I do the correction on a *.jpg version of the grey patch straight from the camera, or do I need to convert from the *.CR2 file?

I expect your JPG is sRGB, so it is not linear. I have said you should use "linear raw" pixels. That means developing the _raw_ image from the CR2 file, and making a _linear_, not sRGB, image.

In addition, the JPG file is 8 bits per channel per pixel, so processing will create banding. I suggest that JPGs are _never_ used for processing if a better source is available.

Your "-evaluate Multiply" string looks okay. Please show complete commands, not just parts of them. I assume you are using "magick", not "magick convert".

Your "-evaluate Add 60" string is bad. What is the text output of "magick -version"? I suspect you have Q16. If so, then values are stored in memory as values between 0 and 65535 (called "QuantumRange"). Adding 60 to such values will have hardly any effect. I suppose you intended 60 in a scale of 0 to 255. So you want 60/255 as a fraction of QuantumRange. Instead of "60", you can use "%[fx:60/255*QuantumRange]". Similarly with "-evaluate Subtract".

When your photo has different shades of gray and you want to make them all neutral, a simple multiplication (a "gain") may not suit all the shades. This is especially true of printed colour charts, when the white card and black ink probably do not have the same spectral reflectances, eg light reflected by black ink contains a greater proportion of blue than light reflected by white card. In those cases, a "gain and bias" may be better: adding and multiplying. These are called "linear" transformations. If we graphed them, the lines would be straight. Instead, we can transform each channel with a polynomial curve. I show this at [dcraw and WB](http://im.snibgo.com/dcrawwb.htm#wbgray)

**Reply by memphis-10** [reply]
Thank you.  I've read over this a few times today and will be working on learning more about everything laid out in this thread next week.  I somewhat assumed correcting jpgs wasn't a good idea.  They were what I had in my virtual machine at the time, so I just went with it for the experiment.  

I'm not at the machine now, but I do have Q16 7.0... I think.  Thank you for pointing out the memory values vs. RGB values.  I read a bit about that difference today and am still wrapping my mind around it.

I have been using "magick convert."  Testing another operation a week or so ago, I tried "convert" only and CMD didn't understand it so I went back to "magick convert" (which I've been using primarily in limited capacity with the scanner/negative project.). I will follow up later with the version and full string.  

Also, thanks for the link to your dcraw and WB page.  I just read through it.  Along with your chart pages, I need some time to experiment and learn.  My current negative inversion script is BAT so I'm somewhat familair, but there is a lot I still need to learn in regard to that language so I can understand what the text is actually doing.  I'll be back at my desk tomorrow, then again next week.

Suppose I have a grey chart in one photo and I'd like it to match it to the grey in another photo--even if the second isn't perfectly grey.  Can I get the values for both and do math to adjust the first to the second, or is it more complicated that that?

Thanks again.

**Reply by fmw42** [reply]
With Imagemagick 7, use magick, not magick convert. For the other tools use magick identify, magick montage, etc. But not magick convert. Magick was completely rewritten to replace convert. But the other tools were not and are just called from magick.

**Reply by memphis-10** [reply]
Thank you.  I've been working to get up to speed with the usage since my original BAT came to me as "magick Convert" but was working so I never questioned it.  Thanks again!

**Comment by snibgo** [comment]
> Suppose I have a grey chart in one photo and I'd like it to match it to the grey in another photo--even if the second isn't perfectly grey. Can I get the values for both and do math to adjust the first to the second, or is it more complicated that that?

It is that simple.

At the risk of making it seem complex, I would explain it like this. Suppose the first photo has a colour patch with RGB values (A, B, C), and the second has a patch (a, b, c). What arithmetic can we do to transform A to a, and B to b, and C to c?

There are many possible transformations. One is to multiply pixels by (a/A, b/B, c/C).

Another is to add values (-A+a, -B+b, -C+c).

Yet another is to raise pixel values to a power ( log(a)/log(A), log(b)/log(B), log(c)/log(C) ).

**Reply by memphis-10** [reply]
Thank you.  I'm a long time film and digital photographer, but very new to the math of digital imaging.  This makes total sense and I will incorporate this lesson in my next round of tests when I revisit the cards after converting to linear TIF.

Is one of the 3 methods you mentioned above better than the other, specifically to avoid or reduce clipping?  The first two are easy to understand with my limited knowledge.  Thanks again!

**Reply by fmw42** [reply]
I mostly use the first method.  multiply pixels by (a/A, b/B, c/C).  See -color-matrix

**Comment by snibgo** [comment]
Strictly speaking, adjusting colour temperature is a special case of the more general white balancing. Colour temperature has one variable: the temperature adjustment. White balancing has two variables: the two independent ratios between the R, G and B multipliers.

When we have linear values, multiplying channels is exactly equivalent to putting a colour filter in front of the lens. Calculating the values and applying the transformation are both easy. When values are not linear but have a gamma transfer function, this still works. sRGB doesn't have an exact gamma transfer function, but it is very close.

Adjusting an image to make one colour patch like another needs one variable, such as multiplying each channel of all pixels by a constant k. Each channel has its own value for k.
```
v' = k * v
```
We can readily calculate k for each channel. This is a "gain" method.

If we have two colour patches to match, we need two variables per channel, such as:
```
v' = k * v + c
```
We can readily calculate k and c. This is "gain and bias", a multiply and an add.

If we have more than two patches to match, we can find the gain-only, or gain-and-bias, that gives the least error. Or we can use a more complex transformation that ensures all patches match exactly.

To avoid clipping with gain-only, ensure that each multiplier is less than one.

For any transformation done with floating-point, we can check if the maximum output is more than 100% of QuantumRange. If it is, divide all values by the maximum.

The power curve is:
```
v' = pow(v, p)
```
This has the property that inputs at zero or one have outputs unchanged, which can be useful. For example, we can adjust what was the mean of an image to be any value we want, such as 50% of QuantumRange overall or in each channel. See "-auto-gamma".

**Reply by memphis-10** [reply]
Thank you.  I've read this a few times and can somewhat digest it.  Rather than ask questions line by line, I'll come back when I get to this point.

I managed to invert my *.CR2 charts with dcraw, 16 bit TIFF, with camera WB.  I experimented some today, but the color balance using the grey chip only wasn't doing what I expected.  I'll work on it more methodically during the week and see if I can figure it out. One thing I noticed is that the image gets darker if I multiply by 1 and lower.  That makes sense, but is there a luminance compensation that can be applied based on the fractional RGB values use to adjust the color?  Again, I'll work on it more, so it can wait until I have solidly documented each step.

Thanks again.  Massive help just learning these little bits.

**Comment by snibgo** [comment]
> ... is there a luminance compensation that can be applied based on the fractional RGB values use to adjust the color?

You could adjust the three multipliers so they average to 1.0, but the result may have pixels values above QuantumRange, so this can cause clipping.

You can divide the result by the maximum value with "-evaluate Divide %[fx:maxima]". This stretch values so the maximum is exactly QuantumRange.

You can use "-auto-level" which stretches values so the minimum is exactly zero and the maximum is exactly QuantumRange.

**Reply by memphis-10** [reply]
Thank you.  I'm working on getting a few charts on negative film as well as practicing the simple math on my dcraw conversions in the meantime.  Just seeing them in raw form is a learning curve after editing in Lightroom the past 10 years.  I truly appreciate the lessons and will go through the basics when I have a good set of images to work with.
